File: packages/live-update/example-ionic-portals/android/app/src/main/java/com/example/plugin/IonicProviderTestPlugin.java
package com.example.plugin;

import androidx.annotation.NonNull;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.ionic.liveupdateprovider.FederatedCapacitorSyncResult;
import io.ionic.liveupdateprovider.LiveUpdateProvider;
import io.ionic.liveupdateprovider.LiveUpdateProviderError;
import io.ionic.liveupdateprovider.LiveUpdateProviderManager;
import io.ionic.liveupdateprovider.LiveUpdateProviderRegistry;
import io.ionic.liveupdateprovider.LiveUpdateProviderSyncCallback;
import io.ionic.liveupdateprovider.LiveUpdateProviderSyncResult;
import java.io.File;
import java.util.HashMap;
import java.util.Map;

/**
 * Capacitor plugin for the example app that exposes the Ionic Live Update
 * Provider SDK to JavaScript so the integration can be tested end-to-end
 * without needing a real Federated Capacitor or Portals host.
 */
@CapacitorPlugin(name = "IonicProviderTest")
public class IonicProviderTestPlugin extends Plugin {

    private static final String PROVIDER_ID = "capawesome";

    @PluginMethod
    public void isProviderRegistered(PluginCall call) {
        LiveUpdateProvider provider = LiveUpdateProviderRegistry.resolve(PROVIDER_ID);
        JSObject ret = new JSObject();
        ret.put("registered", provider != null);
        call.resolve(ret);
    }

    @PluginMethod
    public void getLatestAppDirectory(PluginCall call) {
        try {
            LiveUpdateProviderManager manager = createManagerFromCall(call);
            if (manager == null) {
                return;
            }
            File directory = manager.getLatestAppDirectory();
            JSObject ret = new JSObject();
            ret.put("latestAppDirectory", directory == null ? null : directory.getAbsolutePath());
            call.resolve(ret);
        } catch (Exception exception) {
            call.reject(exception.getMessage() == null ? "Unknown error" : exception.getMessage());
        }
    }

    @PluginMethod
    public void syncManager(PluginCall call) {
        try {
            LiveUpdateProviderManager manager = createManagerFromCall(call);
            if (manager == null) {
                return;
            }
            manager.sync(
                new LiveUpdateProviderSyncCallback() {
                    @Override
                    public void onSuccess(@NonNull LiveUpdateProviderSyncResult result) {
                        JSObject ret = new JSObject();
                        File directory = manager.getLatestAppDirectory();
                        ret.put("latestAppDirectory", directory == null ? null : directory.getAbsolutePath());
                        if (result instanceof FederatedCapacitorSyncResult) {
                            Map<String, Object> metadata = ((FederatedCapacitorSyncResult) result).getMetadata();
                            if (metadata != null) {
                                try {
                                    ret.put("metadata", new JSObject(new org.json.JSONObject(metadata).toString()));
                                } catch (Exception ignored) {
                                    // Best-effort metadata bridging.
                                }
                            }
                        }
                        call.resolve(ret);
                    }

                    @Override
                    public void onFailure(@NonNull LiveUpdateProviderError.SyncFailed error) {
                        call.reject(error.getMessage() == null ? "Sync failed" : error.getMessage());
                    }
                }
            );
        } catch (Exception exception) {
            call.reject(exception.getMessage() == null ? "Unknown error" : exception.getMessage());
        }
    }

    private LiveUpdateProviderManager createManagerFromCall(PluginCall call) {
        String managerKey = call.getString("managerKey");
        if (managerKey == null || managerKey.isEmpty()) {
            call.reject("managerKey must be provided");
            return null;
        }
        LiveUpdateProvider provider = LiveUpdateProviderRegistry.resolve(PROVIDER_ID);
        if (provider == null) {
            call.reject(
                "Provider '" +
                PROVIDER_ID +
                "' is not registered. Make sure capawesomeCapacitorLiveUpdateIncludeIonicProvider = true in variables.gradle."
            );
            return null;
        }
        Map<String, Object> config = new HashMap<>();
        config.put("managerKey", managerKey);
        String appId = call.getString("appId");
        if (appId != null) {
            config.put("appId", appId);
        }
        String channel = call.getString("channel");
        if (channel != null) {
            config.put("channel", channel);
        }
        try {
            return provider.createManager(getContext().getApplicationContext(), config);
        } catch (LiveUpdateProviderError.InvalidConfiguration error) {
            call.reject(error.getMessage() == null ? "Invalid configuration" : error.getMessage());
            return null;
        }
    }
}

File: packages/live-update/example-ionic-portals/ios/App/App/IonicProviderTestPlugin.swift
import Foundation
import Capacitor
import LiveUpdateProvider

/// Capacitor plugin for the example app that exposes the Ionic Live Update
/// Provider SDK to JavaScript so the integration can be tested end-to-end
/// without needing a real Federated Capacitor or Portals host.
@objc(IonicProviderTestPlugin)
public class IonicProviderTestPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "IonicProviderTestPlugin"
    public let jsName = "IonicProviderTest"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isProviderRegistered", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLatestAppDirectory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "syncManager", returnType: CAPPluginReturnPromise)
    ]

    private static let providerId = "capawesome"

    @objc func isProviderRegistered(_ call: CAPPluginCall) {
        let provider = LiveUpdateProviderRegistry.shared.resolve(IonicProviderTestPlugin.providerId)
        call.resolve(["registered": provider != nil])
    }

    @objc func getLatestAppDirectory(_ call: CAPPluginCall) {
        do {
            let manager = try createManager(from: call)
            let path = manager.latestAppDirectory?.path
            call.resolve(["latestAppDirectory": path ?? NSNull()])
        } catch let error as PluginCallError {
            call.reject(error.message)
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func syncManager(_ call: CAPPluginCall) {
        let manager: any LiveUpdateManaging
        do {
            manager = try createManager(from: call)
        } catch let error as PluginCallError {
            call.reject(error.message)
            return
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Task {
            do {
                let result = try await manager.sync()
                var ret: [String: Any] = [:]
                ret["latestAppDirectory"] = manager.latestAppDirectory?.path ?? NSNull()
                if let fedCap = result as? FederatedCapacitorSyncResult, let metadata = fedCap.metadata {
                    ret["metadata"] = metadata
                }
                call.resolve(ret as PluginCallResultData)
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    private func createManager(from call: CAPPluginCall) throws -> any LiveUpdateManaging {
        guard let managerKey = call.getString("managerKey"), !managerKey.isEmpty else {
            throw PluginCallError("managerKey must be provided")
        }
        guard let provider = LiveUpdateProviderRegistry.shared.resolve(IonicProviderTestPlugin.providerId) else {
            throw PluginCallError(
                "Provider '\(IonicProviderTestPlugin.providerId)' is not registered. " +
                "Make sure the IonicProvider subspec is used in the Podfile."
            )
        }
        var config: [String: Any] = ["managerKey": managerKey]
        if let appId = call.getString("appId") {
            config["appId"] = appId
        }
        if let channel = call.getString("channel") {
            config["channel"] = channel
        }
        return try provider.createManager(config: config)
    }
}

/// Internal error type so we can carry a user-facing message back to the
/// `CAPPluginCall.reject` callsite.
private struct PluginCallError: Error {
    let message: String

    init(_ message: String) {
        self.message = message
    }
}
